### PR TITLE
Cache or Clear tables.

### PR DESCRIPTION
**Pull request by harikt**
Hi, 

This PR contains commit from :  https://github.com/cakephp/migrations/pull/308 .

~~This is WIP for there is no tests currently.~~

The idea is allow people to use cakephp/migrations as standalone .

Interested to know if you guys love this addition.

Thank you.

### CODE REVIEW

File: src/Util/SchemaTrait.php [hunk 1]
```diff
@@ -0,0 +1,49 @@
...
+/**
+ * Trait gathering schema collection used for caching
+ */
+trait SchemaTrait
+{
+
+    /**
+     * Helper method to get the schema collection.
+     *
+     * @return false|\Cake\Database\Schema\Collection
```
Comment (dereuromark) [inline src/Util/SchemaTrait.php]: The fallback value should be null for objects, not false.

Comment (harikt) [inline src/Util/SchemaTrait.php]: sure. I can make the change.

**Comment by ADmad** [comment]
Looks like stickler is using the wrong standard for checks. The `.stickler.yml` from cakephp/cakephp can be copied.

**Comment by lorenzo** [comment]
It does make sense to me to move those shells to this plugin

**Comment by harikt** [comment]
Hey guys, 

Can anyone help me finish writing tests for this ?

```
vendor/bin/phpunit tests/TestCase/Command/CacheBuildTest.php
```

throws

```
1) Migrations\Test\Command\CacheBuildTest::testExecute
Symfony\Component\Console\Exception\InvalidArgumentException: The "source" option does not exist.
```

Thank you.

**Comment by lorenzo** [comment]
Seems like you are passing the `source` CLI option somewhere in your test?

**Comment by harikt** [comment]
@lorenzo source ? No. See the changes. Some how the tests are failing :-/ .

File: tests/TestCase/Command/CacheBuildTest.php [hunk 2]
```diff
@@ -53,8 +50,10 @@ class CacheBuildTest extends TestCase
     public function setUp()
     {
         parent::setUp();
-
+        Cache::enable();
         $this->connection = ConnectionManager::get('test');
+        $this->connection->cacheMetadata(true);
```
Comment (lorenzo) [inline tests/TestCase/Command/CacheBuildTest.php]: You should revert this in `tearDown`

File: tests/TestCase/Command/CacheBuildTest.php [hunk 3]
```diff
@@ -53,8 +50,10 @@ class CacheBuildTest extends TestCase
     public function setUp()
     {
         parent::setUp();
-
+        Cache::enable();
```
Comment (lorenzo) [inline tests/TestCase/Command/CacheBuildTest.php]: You need to disable in `tearDown`

**Comment by lorenzo** [comment]
@harikt ok, looks like it is failing because you are missing some `tearDown` logic

**Comment by harikt** [comment]
@lorenzo other tests seems failing :( . I will investigate.

**Comment by lorenzo** [comment]
@harikt they fail because of corrupted state in a previous test

**Comment by harikt** [comment]
So there is a problem with the serialized value and the one running on containers. Do you know how we can fix that ?

Also there is a permission issues I noticed for some tests to fail on localhost. Files are in `/tmp/` folder.

```
vendor/bin/phpunit 
PHPUnit 4.8.35 by Sebastian Bergmann and contributors.

.............................S.................................SS 65 / 99 ( 65%)
S..........................EEEEEEE

Time: 5.02 seconds, Memory: 18.00MB

There were 7 errors:

1) Migrations\Test\MigrationHelperTest::testIndexMethod
dir(/tmp/systemd-private-0645317efcd547448e31d26ec6899e87-rtkit-daemon.service-BskPka/): failed to open dir: Permission denied

/var/www/projects/github.com/harikt/cakephp-orm-migrations/vendor/cakephp/cakephp/src/Cache/Engine/FileEngine.php:309
/var/www/projects/github.com/harikt/cakephp-orm-migrations/vendor/cakephp/cakephp/src/Cache/Engine/FileEngine.php:286
/var/www/projects/github.com/harikt/cakephp-orm-migrations/vendor/cakephp/cakephp/src/Cache/Cache.php:451
/var/www/projects/github.com/harikt/cakephp-orm-migrations/tests/TestCase/View/Helper/MigrationHelperTest.php:46

2) Migrations\Test\MigrationHelperTest::testColumnMethod
dir(/tmp/systemd-private-0645317efcd547448e31d26ec6899e87-rtkit-daemon.service-BskPka/): failed to open dir: Permission denied

/var/www/projects/github.com/harikt/cakephp-orm-migrations/vendor/cakephp/cakephp/src/Cache/Engine/FileEngine.php:309
/var/www/projects/github.com/harikt/cakephp-orm-migrations/vendor/cakephp/cakephp/src/Cache/Engine/FileEngine.php:286
/var/www/projects/github.com/harikt/cakephp-orm-migrations/vendor/cakephp/cakephp/src/Cache/Cache.php:451
/var/www/projects/github.com/harikt/cakephp-orm-migrations/tests/TestCase/View/Helper/MigrationHelperTest.php:46

3) Migrations\Test\MigrationHelperTest::testColumns
dir(/tmp/systemd-private-0645317efcd547448e31d26ec6899e87-rtkit-daemon.service-BskPka/): failed to open dir: Permission denied

/var/www/projects/github.com/harikt/cakephp-orm-migrations/vendor/cakephp/cakephp/src/Cache/Engine/FileEngine.php:309
/var/www/projects/github.com/harikt/cakephp-orm-migrations/vendor/cakephp/cakephp/src/Cache/Engine/FileEngine.php:286
/var/www/projects/github.com/harikt/cakephp-orm-migrations/vendor/cakephp/cakephp/src/Cache/Cache.php:451
/var/www/projects/github.com/harikt/cakephp-orm-migrations/tests/TestCase/View/Helper/MigrationHelperTest.php:46

4) Migrations\Test\MigrationHelperTest::testColumn
dir(/tmp/systemd-private-0645317efcd547448e31d26ec6899e87-rtkit-daemon.service-BskPka/): failed to open dir: Permission denied

/var/www/projects/github.com/harikt/cakephp-orm-migrations/vendor/cakephp/cakephp/src/Cache/Engine/FileEngine.php:309
/var/www/projects/github.com/harikt/cakephp-orm-migrations/vendor/cakephp/cakephp/src/Cache/Engine/FileEngine.php:286
/var/www/projects/github.com/harikt/cakephp-orm-migrations/vendor/cakephp/cakephp/src/Cache/Cache.php:451
/var/www/projects/github.com/harikt/cakephp-orm-migrations/tests/TestCase/View/Helper/MigrationHelperTest.php:46

5) Migrations\Test\MigrationHelperTest::testValue
dir(/tmp/systemd-private-0645317efcd547448e31d26ec6899e87-rtkit-daemon.service-BskPka/): failed to open dir: Permission denied

/var/www/projects/github.com/harikt/cakephp-orm-migrations/vendor/cakephp/cakephp/src/Cache/Engine/FileEngine.php:309
/var/www/projects/github.com/harikt/cakephp-orm-migrations/vendor/cakephp/cakephp/src/Cache/Engine/FileEngine.php:286
/var/www/projects/github.com/harikt/cakephp-orm-migrations/vendor/cakephp/cakephp/src/Cache/Cache.php:451
/var/www/projects/github.com/harikt/cakephp-orm-migrations/tests/TestCase/View/Helper/MigrationHelperTest.php:46

6) Migrations\Test\MigrationHelperTest::testAttributes
dir(/tmp/systemd-private-0645317efcd547448e31d26ec6899e87-rtkit-daemon.service-BskPka/): failed to open dir: Permission denied

/var/www/projects/github.com/harikt/cakephp-orm-migrations/vendor/cakephp/cakephp/src/Cache/Engine/FileEngine.php:309
/var/www/projects/github.com/harikt/cakephp-orm-migrations/vendor/cakephp/cakephp/src/Cache/Engine/FileEngine.php:286
/var/www/projects/github.com/harikt/cakephp-orm-migrations/vendor/cakephp/cakephp/src/Cache/Cache.php:451
/var/www/projects/github.com/harikt/cakephp-orm-migrations/tests/TestCase/View/Helper/MigrationHelperTest.php:46

7) Migrations\Test\MigrationHelperTest::testStringifyList
dir(/tmp/systemd-private-0645317efcd547448e31d26ec6899e87-rtkit-daemon.service-BskPka/): failed to open dir: Permission denied

/var/www/projects/github.com/harikt/cakephp-orm-migrations/vendor/cakephp/cakephp/src/Cache/Engine/FileEngine.php:309
/var/www/projects/github.com/harikt/cakephp-orm-migrations/vendor/cakephp/cakephp/src/Cache/Engine/FileEngine.php:286
/var/www/projects/github.com/harikt/cakephp-orm-migrations/vendor/cakephp/cakephp/src/Cache/Cache.php:451
/var/www/projects/github.com/harikt/cakephp-orm-migrations/tests/TestCase/View/Helper/MigrationHelperTest.php:46

FAILURES!
Tests: 99, Assertions: 328, Errors: 7, Skipped: 4.
```

If we do a `git status` there are many files added. Seems we could add those to gitignore itself.

```
 ✘  /var/www/projects/github.com/harikt/cakephp-orm-migrations   cache  gs
On branch cache
Your branch is up-to-date with 'origin/cache'.
Untracked files:
  (use "git add <file>..." to include in what will be committed)

	cakephp_test
	composer.lock
	tests/test_app/Plugin/TestBlog/config/Migrations/
	tests/test_app/config/Create/20170323150809_TestCreate.php
	tests/test_app/config/Migrations/20170323150811_TestAutoIdDisabledSnapshotSqlite.php
	tests/test_app/config/Migrations/20170323150811_TestNotEmptySnapshotNoLockSqlite.php
	tests/test_app/config/Migrations/20170323150811_TestNotEmptySnapshotSqlite.php
	tests/test_app/config/Migrations/20170323150813_CreateUsers.php
	tests/test_app/config/TestsMigrations/schema-dump-test.lock

nothing added to commit but untracked files present (use "git add" to track)
```

I am leaving this to rest for now. Will come back when I have some more energy :) .

**Comment by harikt** [comment]
Closing this PR. Opening a rebased one.

**Pull request closed without merging**